File: medium/Linked-List-Random-Node_medium/Solution1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
 import java.util.Random;
public class Solution {

    ListNode head;
    Random ran;
    /** @param head The linked list's head. Note that the head is guanranteed to be not null, 
    so it contains at least one node. */
    public Solution(ListNode head) {
        this.head = head;
        ran = new Random();
    }
    
    /** Returns a random node's value. */
    public int getRandom() {
        ListNode cur = head;
        ListNode res = head;
        for(int n=1; cur!= null; n++) {
            if(ran.nextInt(n) == 0) {
                res = cur;
            }
            cur = cur.next;
        }
        return res.val;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */

File: algorithms/Design_HashMap-easy/solution1.cpp
class MyHashMap {
    vector<list<pair<int, int>>> map;
    int cap;
public:
    MyHashMap() {
        cap = 2069;
        map.resize(2069);
    }
    
    void put(int key, int value) {
        int index = key % cap;
        list<pair<int, int>>& cur = map[index];
        for (auto& p: cur) {
            if (p.first == key) {
                p.second = value;
                return;
            }
        }
        cur.push_back({key, value});
    }
    
    int get(int key) {
        int index = key % cap;
        list<pair<int, int>>& cur = map[index];
        for (auto& p: cur) {
            if (p.first == key) {
                return p.second;
            }
        }
        return -1;
    }
    
    void remove(int key) {
        int index = key % cap;
        list<pair<int, int>>& cur = map[index];
        for (auto& p: cur) {
            if (p.first == key) {
                cur.remove(p);
                break;
            }
        }        
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */
